### ISSUE TITLE
upgraded to 4.2 - icons are not sharp

### ISSUE THREAD

**Issue by morrow95** [FortAwesome/Font-Awesome#4296]
Just upgraded to 4.2 from 4.1... nothing has changed except the version.  Right off the bat I noticed the icons are not as sharp as before - they are somewhat pixelated.  What was changed between these two versions to cause this?

**Comment by rjgotten** [comment]
@tagliala

On browsers rendering with DirectWrite, the default behavior is to allow full subpixel precision rendering of a font at sizes of 21px or greater. And the icons in the font set look *fine* at those sizes, if occasionally somewhat blurred.

The problem is that at pixel sizes below 21px, browsers apply the hinting embedded in the font, which in FontAwesome 4.2 and up is just utterly, utterly broken. This problem is **not limited to brand icons**. It affects all the icons in the set, but is just more noticible on some icons than on others.

You can, infact, work around the poor rendering fidelity due to the broken hinting by setting a css rotation small enough to remain visually undectable, but large enough to force the affected browsers (IE, Firefox, Chrome) to start rendering fonts with full sub-pixel geometric precision.

Here's a snap of the **unfixed** icon in the searchbox on FontAwesome's own icon overview page, as seen in Firefox on Windows 7 with Direct2D and DirectWrite enabled:
![Unfixed](http://s28.postimg.org/xiwz97s4t/no_fix.png)

And here's the same, with a `transform:rotate(0.001deg)` added to the `.fa` class via the developer tools:
![Fixed](http://s28.postimg.org/x4vn9m819/fix.png)

Notice the difference there?

**Comment by tagliala** [comment]
@rjgotten thanks for the very detailed info

> The problem is that at pixel sizes below 21px, browsers apply the hinting embedded in the font, which in FontAwesome 4.2 and up is just utterly, utterly broken. This problem is not limited to brand icons. It affects all the icons in the set, but is just more noticible on icons with intricate detail (like most brand icons...) than on others.

any info on how Dave could fix this?

**Comment by rjgotten** [comment]
> any info on how Dave could fix this?

You need the proper hinting set up to get the font to render at subpixel position. Iirc if hinting is not present, then both DirectWrite and GDI ClearType will default to snapping towards the nearest whole pixel. Afaict this is what FontAwesome 4.2+ is currently doing.

Browsers rendering with DirectWrite have a hardcoded behavior to skip hinting tables altogether and render at full subpixel precision for font-sizes greater than 20 pixels. The rotation hack forces them to do the same with font-sizes of 20 pixels or smaller. However, that does not fix the issue for browsers stuck with GDI ClearType, which iirc does not have the same behavior of starting to render at full subpixel precision.

I don't think there's a way around it: you *must* have proper hinting tables to fix this completely.
The good news is that having TTFAutoHint generate hinting works OK for most icons, atleast given a cursory glance.

**Comment by tagliala** [comment]
@davegandy

**Comment by SiteSplat** [comment]
Yup applying this to 21px and below is a temp fix: 

transform:rotate(0.001deg)

Kudos to @rjgotten 

Somone pointed out this issue before (about a year ago) but it was dismissed just by saying "use multiplier of 14px"

>This has very little to do with rendering at exact multiples of 14px. (Which is a big WTF, btw.)

@tagliala @davegandy  Taking the time to verify the issues rather than dismissing it with no basis would be more productive considering other people take time to file these issues to help out in the first place...


Dave

**Comment by tagliala** [comment]
@SiteSplat 

https://github.com/FortAwesome/Font-Awesome/issues/4296#issuecomment-54428735 I took time to verify :)

**Comment by SiteSplat** [comment]
Ok, has a solution been found yet? this issue has been around for well over a year now.  I'm not sure why its labeled as "enhancement" when its a bug. It did not happen until after v 4.1.0.

 If these things are not addressed right away what's the point of even releasing new versions? Needless to say that a prompted reaction for a product of this importance is the first thing should be in place.

I also have concern about the black tie package at this point...

**Comment by neuropass** [comment]
ping @tagliala @davegandy has this been addressed on v 4.5?  I dont see it in the changelog. Thanks!

**Comment by tagliala** [comment]
please provide evidence and affected browsers for this issue

**Comment by SiteSplat** [comment]
Naaa why would they address it? 20 new icons are more important then fixing a core bug that 40k++++ people  are stuck with. 

Do your self a favor and start using SVGs or another font library instead. This has been opened for a year and now apparently closed up as well. How ridiculous 

@tagliala Provide evidence? are you reading the tickets at all?

**Comment by tagliala** [comment]
@SiteSplat this issue has 19 comments out of 40k+++ people and 8 comments are mine.

I cannot reproduce this bug, as I said multiple times.

Please provide a couple of fiddles that show this issue. Please also include browsers and OSes affected

**Comment by rjgotten** [comment]
@tagliala 

> I cannot reproduce this bug, as I said multiple times.

"Works on my machine" is not a valid reason to close a bug, no matter which way you slice it. That's just bad form.

> please provide evidence and affected browsers for this issue, meanwhile closing here

There is no need to ask for further proof of the issue's existence, when you already have the screen grabs I lifted straight from the official FontAwesome website with and without the temporary workaround that I proposed applied.

Those list Firefox with Direct2D and DirectWrite enabled on Windows as the target browser, platform and rendering technologies. But it's the same for Chrome and Internet Explorer which also use Direct2D and DirectWrite. (See the pattern there?)

You even asked right here in this thread how to go about fixing this problem and I described the process. The fact that you even asked for how to fix this means you already acknowledged the problem. Backing out now quite frankly makes you look like a *total tool*.

**Comment by SiteSplat** [comment]
@rjgotten I couldn't have said it better.  Funny because I was just gonna quote your posts from Sep 9th and 11th. Then I thought, why bother? They obviously don't want to fix it. I had pointed this out just a few days after the original poster: https://github.com/FortAwesome/Font-Awesome/issues/4442

**Comment by tagliala** [comment]
@rjgotten 

>  They obviously don't want to fix it.

This is far from being true.

@rjgotten

> "Works on my machine" is not a valid reason to close a bug, no matter which way you slice it. That's just bad form.

I could say the same for your machine.

Just tested this stuff again on 3 machines and 1 smartphone, here are the results

## Win 10
:white_check_mark: Chrome 46 / Win 10
:white_check_mark: MS Edge 13 / Win 10
:white_check_mark: Firefox 42 / Win 10
:white_check_mark: IE 11 / Win 10
![immagine2](https://cloud.githubusercontent.com/assets/556268/11695324/d1a8e854-9eae-11e5-9d0d-cac2ea65800b.png)

## OSX 10.11.2 (non-retina)
:white_check_mark:  Chrome 46 / OSX 10.11.1
:white_check_mark: Firefox 42 / OSX 10.11.1
:white_check_mark:  Safari 9.0.1 / OSX 10.11.1
![screen shot 2015-12-09 at 20 00 17](https://cloud.githubusercontent.com/assets/556268/11695619/78a78ab0-9eb0-11e5-9807-7b1c23caa4d7.png)

# Win 7 (vm box)
:white_check_mark:  Chrome 46 / Win 7
:x: Firefox 42 / Win 7
:x:  IE 9 / Win 7
![immagine3](https://cloud.githubusercontent.com/assets/556268/11695327/d4a0a59c-9eae-11e5-9c7a-95f327a26af4.png)

Please also notice this issue in FF42 / Windows 7:
![screen shot 2015-12-09 at 20 07 58](https://cloud.githubusercontent.com/assets/556268/11695690/db78035e-9eb0-11e5-895f-1d48476b85bd.png)

As I said multiple times, Dave had to change something to address #3683 and #3440 and that bug affected more browsers and OS of this one, even at the base font size of 14px. 

I've also tested glyphicons (http://jsfiddle.net/tagliala/9qg96Lna/2/):
![immagine4](https://cloud.githubusercontent.com/assets/556268/11696168/4a17c07c-9eb3-11e5-87e3-9ca13a6b80cb.png)

As you can see, they have the same issues of FA, even at the base font size of 16px. They work better at 18px. You may consider switching to glyphicons for your use case

AFAIK, Dave uses Font Squirrel to generate webfonts (http://www.fontsquirrel.com/tools/webfont-generator) giving the .otf file as input. If you are able to find settings which improve the rendering (without messing up with #3683 and #3440), feel free to share

The title of this issue is: "upgraded to 4.2 - icons are not sharp" and this is not true. They are not sharp at non multiples of 14px on some browser (FF / IE) and some OSes (win7, maybe 8?). In this case I will reopen for sure, but do not expect a magic wand.

Hope it helps

**Comment by rjgotten** [comment]
> Please also notice this issue in FF42 / Windows 7:

Uhm... yeah. That could very well be because Firefox, when run in most VMs *disables* DirectWrite and falls back to GDI rendering. The default behavior of GDI rendering in Windows applications is to adher to the OS setting for ClearType. And in most VM installs it will be *turned off* by default.

Hence all the jagged edges in Firefox on both the icon font and the normal text.

IE though, while it will disable DirectWrite, will by default override the OS level setting for ClearType and will render with ClearType *enabled*, so in IE's case the normal text will be ok, and only the icon font will be jaggy because the icon font is not properly hinted.

Chrome will, when DirectWrite is not available, fall back to its own internal font renderer which, if I remember correctly, will render with full geometric precision when no hinting is present.


So.. I guess you just delivered the evidence you were looking for that proves the hinting is broken?

**Comment by Luukyb** [comment]
+1
We are having the same issue here, IE11 on Win 7, search icon at 16px looks pixelated.

**Comment by neuropass** [comment]
Closed becuase? How about "WONT FIX IT" ? looks like more appropriate.

**Comment by hakimio** [comment]
rjgotten's proposed workaround works, but the issue should NOT be closed because it has never been fixed and use "multiple's of 14px" is just a plain stupid suggestion.

**Comment by DaleGardner** [comment]
Thank you @rjgotten ! :clap: 

Here is what the icon looks like at 14px on Windows 10 x64 with Firefox Developer Edition x64
![image](https://cloud.githubusercontent.com/assets/3970939/15376172/5c7e3556-1d21-11e6-8059-301edaa0553d.png)

And with @rjgotten workaround applied:
![image](https://cloud.githubusercontent.com/assets/3970939/15376242/c14b7d9a-1d21-11e6-8dd0-78abadc97893.png)

**Comment by dnajar2** [comment]
I know this topic has been closed out I just want to add that if you need to resize FA font size use "em" instead of pixels and it looks just fine in Firefox  

resized font using pixels
 
![pasted image at 2016_06_16 09_45 am](https://cloud.githubusercontent.com/assets/15351543/16155394/0e4e1680-3464-11e6-9cc3-89ff8873b865.png)

Resized font using 'em'
![pasted image at 2016_06_16 10_03 am](https://cloud.githubusercontent.com/assets/15351543/16155405/1c58a952-3464-11e6-867e-6fe8fbd6ff68.png)

**Comment by hakimio** [comment]
Or just use rjgotten's fix to make it work with pixels as well:
`transform: rotate(0.001deg)`

**Comment by rjgotten** [comment]
> I know this topic has been closed out I just want to add that if you need to resize FA font size use "em" instead of pixels and it looks just fine in Firefox 

Your example is rendering the coinstack icon at a larger size in the 'em' example than it is using in the 'px' example. That larger size just happens to produce a decent icon. Might well be that the larger icon's computed size exceeds 20 ~ 21 px and thus the icon would start rendering with full geometric precision.

**Comment by SiteSplat** [comment]
yeah no @dnajar2 that is not a fix... I 'm not sure why they are ignoring this. Its quite embarrassing if yo uask me for a project this popular. Not even answering to this numerous posts @davegandy . Really? then we add PLEASE HELP US txt file everywhere to support a project? Just get a grip.  Can you also remove this from the main font awesome page:

https://gyazo.com/e52acfc3306d8cf90472a632837582c3

Unless you fix this ticket that statement about font awesome is a straight lie.

**Comment by dnajar2** [comment]
@SiteSplat all I was saying is that using em's worked for me, I was having a bad day with F.F. and when asked to take an approached that would fix the bad rendering of the fa-database icon I simply used em's to resize the font, because fa-2x was not big enough and fa-3x was too big, didn't mean to rattle a closed cage.

**Comment by davegandy** [comment]
So for reference, this has always bugged me like crazy. But it's less at the top of mind as I don't work on windows. I'm going to take a crack at fixing this.

**Comment by tagliala** [comment]
@davegandy please check that #3683 and #3440 will not come back when you are trying to address this bug.

I could add http://jsfiddle.net/tagliala/1rf32u7y/ and http://jsfiddle.net/24Mry/2/ to visual test cases, let me know

**Comment by edoher** [comment]
I ran into this lack of sharpness issue today, while using FA 4.3. I updated the Fiddle here [http://jsfiddle.net/1rf32u7y/2/](url) to show this problem, comparing both the rotated and non rotated versions of the icons. The search icon is clearly much sharper when using the rotation fix. Will this ever be fixed?

(Screenshot using Firefox 47.0.1 in Windows 7.)

![2016-10-19_1127](https://cloud.githubusercontent.com/assets/12150586/19513178/5a4dbaf8-95ef-11e6-8aac-607bbc53dd2c.png)

**Comment by kodecount** [comment]
Thanks @rjgotten for the quick fix till it is completely fixed.

**Comment by tompazourek** [comment]
I'm not sure if it's exactly the same issue, but I visited this page on my PC:

http://fontawesome.io/icon/id-card-o/

Looking into the smallest 14px icon, it seem to appear blurry sometimes.

When I resize the window horizontally, it's sometimes crisp and sometimes blurry. It looks like some ClearType issue or something.

These are pictures of the same icon, just with the browser window having different widths:

![capture](https://cloud.githubusercontent.com/assets/370699/23993898/c6777b3e-0a42-11e7-83b6-ef58ce1182a5.PNG) ![capture2](https://cloud.githubusercontent.com/assets/370699/23993899/c691735e-0a42-11e7-927f-4d10fe78316d.PNG)

Using Windows 10 and Chrome 56.

**What is happening? Can someone explain?**

**Comment by tagliala** [comment]
@tompazourek nope, it is not the same issue.

It depends on centered content, and it is present on FA 3.2.1 too: https://jsfiddle.net/tagliala/veL8gdxx/1/

It is more related to #4306 than this one. Other icon fonts should be also affected.

**Comment by tompazourek** [comment]
@tagliala Sorry for being off-topic here then. Thanks for pointing me to #4306.

**Comment by tagliala** [comment]
@tompazourek no problem, thanks for bringing back that issue

**Comment by UncleGoogle** [comment]
To apply rotate workaround to element that already have transform you can use wrapper element or for example `before`:

    .elementAlreadyHavingTransition:before {
        transform: rotate(0.001deg)
    }

**Comment by BrainFeeder** [comment]
This issue seems to be around for a while now. I remember me using an older version again in the past because I was stressing out on the jaggered edges.

Now, today, we purchased the latest PRO version. And guess what... Jaggered edges on smaller icons. I've seen the advice to only use multiples of 14px, or 16px (since the base font-size got bumped up.. WTH happened to 14 is the new 12???) but that ain't going to happen over here!

We are using the icons in a CMS, a CRM, all the websites we build... These are not mobile-only applications. That said, we use the icons in combination with Bootstrap 4. If we put an icon in a tab, .btn-sm or between <small> tags the font turns ugly for most of the icons. And we can't just go and change all font-sizes for those icons. 

If the rotate fix helps, you should obviously insert it in all stylesheets! I hope this will be in the next release or we'll be one of those paying users that'll stop paying.

**Comment by tagliala** [comment]
@BrainFeeder

> WTH happened to 14 is the new 12???

Yes, kinda. The base font size of FA5 is 16px, so you will have issues with all other sizes that aren't multiples of 16px. If you had issues with FA4 and 12px, you will have issues with FA5 and 14px

> Now, today, we purchased the latest PRO version.

Please open a new request in the Pro version repository. If you don't have access to that repo, please make sure you have linked your github account at fontawesome.com (in your user settings)

**Comment by sensibleworld** [comment]
Thanks for opening this issue! Since it’s been a while and things have changed quite a bit in Font Awesome since this was originally posted, we’re closing it out to help keep our issue tracker focused on current work.

If this is still important for your project, please feel free to re-open or file a fresh issue with updated details. Thanks so much!
